Add GetLinked query

jackscotti opened the pull request:
Get a list of the content which has a link of the given type to the given
content-id. This is the reverse lookup from the "get-links" lookup - find the
things which link to a piece of content.

Ticket:
https://trello.com/c/zppxFP6p/345-implement-the-publisher-get-linked-endpoint-and-refactor-how-links-are-stored-in-publishing-api

tijmenb commented on app/controllers/v2/link_sets_controller.rb:
```diff
@@ -20,5 +28,9 @@ def links_params
       params.fetch(:content_id)
     end
+
+    def fields
```
Would it be clearer to inline this?

tijmenb commented on app/queries/get_linked.rb:
```diff
@@ -0,0 +1,51 @@
+                      .joins(:link_set)
+                      .where(target_content_id: target_content_id, link_type: link_type)
+
+      content_items = DraftContentItem.where(content_id: content_ids)
```
This doesn't `select` the columns yet.

tijmenb commented on app/controllers/v2/link_sets_controller.rb:
```diff
@@ -12,6 +12,14 @@ def put_links
+    def get_linked
+      render json: Queries::GetLinked.new(
+        content_id: content_id,
+        link_type: link_type,
```
This `link_type` isn't defined anywhere so the endpoint crashes. We should add integration tests that cover the controller.

jackscotti commented:
@tijmenb updated

tijmenb commented on app/queries/get_linked.rb:
```diff
@@ -0,0 +1,66 @@
+      draft_items = DraftContentItem.where(content_id: content_ids)
+
+      live_items_without_draft = LiveContentItem
+                                  .where(content_id: content_ids)
```
I think https://github.com/alphagov/publishing-api/pull/158 applies here too - could you check?

tijmenb commented on app/queries/get_linked.rb:
```diff
@@ -0,0 +1,66 @@
+                      .joins(:link_set)
+                      .where(target_content_id: target_content_id, link_type: link_type)
+
+      draft_items = DraftContentItem.where(content_id: content_ids)
```
I think https://github.com/alphagov/publishing-api/pull/158 applies here too - could you check?

jackscotti replied:
I ran `get_linked_spec` checking the logs with and without preloading the `LiveContentItems`. The queries to the database are exactly the same.

tijmenb replied:
It still does, unfortunately:

https://github.com/alphagov/publishing-api/blob/add-get-linked-endpoint/app/queries/get_linked.rb#L17

You are requesting `content_item.published?` for each draft content item you're requesting here, which will load a LiveContentItem each time. An `includes(:live_content_item)` will fix that.

tijmenb commented:
Looks good broadly. Last things:

- We need to add docs to https://github.com/alphagov/publishing-api/blob/master/doc/publishing-api-syntactic-usage.md
- I'm sure sure about `/linked/:link_type/:content_id`. Maybe `/linked/:content_id/:link_type` is more in line with the other methods? Or perhaps make move `type` to a parameter (`GET /linked/:content_id?link_type=topics&fields[]=blabla`). That way it also looks more like the others.
- We have to add the endpoint to `gds-api-adapters`. You can check https://github.com/alphagov/gds-api-adapters/commit/3aa6816a3846bac6d0c9b5d8799b102766b1c1a3 for an example, it shouldn't be too far off. You'll have to add some "provider state" too in `pact_helper` ([see]( https://github.com/alphagov/publishing-api/blob/8c43a6ef8bd660ecbb7e9c796ca3b6fcfaa8eee5/spec/service_consumers/pact_helper.rb#L252-L257)).

jackscotti commented:
`link_type` is compulsory, I have the feeling it should be present in the query url, maybe as:

`GET /linked/:content_id/link_type/:link_type?fields[]=balblabla`

What do you think?

jackscotti commented:
I've updated the query as you suggested. 

I will now proceed in this way:
- get this PR merged
- update gds-api-adapters
- add the documentation to the endpoint

I need to proceed in this order because when writing the doc I will need to link to the "Request/Response detail" that will be added when updating gds-api-adapters.

tijmenb commented on spec/controllers/v2/link_sets_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+
+RSpec.describe V2::LinkSetsController do
+  let(:content_id) { SecureRandom.uuid }
+  let(:link_type) { "topic" }
```
These `let`'s don't add much, because they're only used once per test. Maybe clearer to inline them?

jackscotti replied:
I removed the `link_type` let. The other one is actually used.

jackscotti commented:
Comments addressed. Thanks for the feedback and for showing me how to find out a N+1 query!

tuzz commented on app/controllers/v2/link_sets_controller.rb:
```diff
@@ -12,6 +12,14 @@ def put_links
+        link_type: params.fetch(:link_type),
+        fields: params.fetch(:fields),
+      ).call
+    end
```
If you don't provide a `link_type` or `fields` parameter, would you get a 500? Can we make this a 4xx response with a descriptive message?

jackscotti replied:
you are right, going to do it. Can you advise where I should return the 4xx response? Here in the controller or in the query itself?

I'd say here.

tuzz commented on app/queries/get_linked.rb:
```diff
@@ -0,0 +1,66 @@
+                                  .where.not(content_id: draft_items.map(&:content_id))
+
+      draft_items + live_items_without_draft
+    end
```
I wonder if some of this could be pushed down to methods/scopes in the models. We might want to look up all link sets that have links that reference some content_id, for example, in other places in the app.

If you did this, there'd be an opportunity to give some descriptive names to these operations, too – rather than having to make sense of these queries to decipher what their "meaning" is.

It's totally up to you and we can refactor these into the models if/when we need to reuse them. Just a thought.

tuzz commented on spec/queries/get_linked_spec.rb:
```diff
@@ -0,0 +1,99 @@
+
+RSpec.describe Queries::GetLinked do
+  content_id = SecureRandom.uuid
+  target_content_id = SecureRandom.uuid
```
Please could `#let` these. It helps to mitigate the risk of tests interfering with each other.

tuzz commented:
Generally looks good :+1: – a few minor points that are probably more suggestions rather than directives.

jackscotti commented:
@tuzz, updated the controller to return a meaningful message when `link_type` or `fields` parameters are not provided. I've also updated the `let`s.

I would like to get this merged first and do the refactoring subsequently in a separate PR. I'd also like to refactor the app to rescue the missing parameters in the application controller rather than only in the linkset controller.
